### PR TITLE
Update lowering llvm 15

### PR DESCRIPTION
**Pull request by lkorenc**
This PR adds a few features related to lowering:
 * Variable initialisation
 * Preliminary work to support GEPs and ABI lowering
 * Floating point constants
 * A few more simple patterns for `hl -> llvm`
 * Minor fixes

Also extra tests are added and the whole pipeline is updated to work with llvm-15.

### CODE REVIEW

File: include/vast/Dialect/LowLevel/LowLevel.td [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+include "mlir/IR/OpBase.td"
+include "mlir/IR/OpAsmInterface.td"
+include "mlir/IR/SymbolInterfaces.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+include "mlir/Interfaces/InferTypeOpInterface.td"
+include "mlir/IR/RegionKindInterface.td"
+
+include "vast/Interfaces/SymbolInterface.td"
+
+def LowLevel_Dialect : Dialect {
```
Comment (xlauko) [inline include/vast/Dialect/LowLevel/LowLevel.td:15]: Add some overview to main readme -- mention `ll` dialect in Dialects section.

File: include/vast/Dialect/LowLevel/LowLevel.td [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    let name = "ll";
+    let summary = "A vast low-level dialect.";
+    let description = [{
+        This dialect is supposed to serve as bottom layer
+        in vast dialect stack. By default there should always
```
Comment (xlauko) [inline include/vast/Dialect/LowLevel/LowLevel.td:20]: This dialect serves as a bottom layer
in VAST dialect tower.

File: include/vast/Dialect/LowLevel/LowLevelOps.td [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    , Results<(outs AnyType:$result)>
+{
+    let summary = "Declaration of variable that have not been initialized yet.";
+    let description = [{ Declaration of variable that have not been initialized yet. }];
+
+    let results = (outs AnyType:$result);
+}
+
+def InitializeVar
+    : LowLevel_Op< "initialize", [NoSideEffect] >
```
Comment (xlauko) [inline include/vast/Dialect/LowLevel/LowLevelOps.td:30]: cannot this have side effects?

File: include/vast/Dialect/LowLevel/LowLevelOps.td [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+#ifndef VAST_DIALECT_IR_LOWLEVELOPS
+#define VAST_DIALECT_IR_LOWLEVELOPS
+
+include "mlir/Interfaces/CallInterfaces.td"
+include "mlir/IR/BuiltinAttributes.td"
+include "vast/Interfaces/SymbolInterface.td"
+
+def LowLevel_StructGEPOp
+  : LowLevel_Op< "gep" >
```
Comment (xlauko) [inline include/vast/Dialect/LowLevel/LowLevelOps.td:11]: [NoSideEffect]

File: include/vast/Dialect/LowLevel/LowLevelOps.td [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  , Arguments<(ins AnyType:$record, I32Attr:$idx, StrAttr:$name)>
```
Comment (xlauko) [inline include/vast/Dialect/LowLevel/LowLevelOps.td:12]: This should be some PointerLikeType or ObjectLikeType, at least add TODO to fix later.
Same for result type?

File: include/vast/Dialect/LowLevel/Passes.td [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) 2022-present, Trail of Bits, Inc.
+
+#ifndef VAST_DIALECT_LOWLEVEL_PASSES_TD
+#define VAST_DIALECT_LOWLEVEL_PASSES_TD
+
+include "mlir/Pass/PassBase.td"
+
+def ToLLVM : Pass<"vast-ll-to-llvm", "mlir::ModuleOp"> {
```
Comment (xlauko) [inline include/vast/Dialect/LowLevel/Passes.td:8]: Rename to LLToLLVM

File: include/vast/Util/DialectConversion.hpp [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    struct BasePattern : mlir::OpConversionPattern< Op >
+    {
+        using parent_t = mlir::OpConversionPattern< Op >;
+        using operation_t = Op;
+        using parent_t::parent_t;
+
+        mlir::LogicalResult matchAndRewrite(
+                operation_t op,
+                typename operation_t::Adaptor ops,
+                mlir::ConversionPatternRewriter &rewriter) const override
```
Comment (xlauko) [inline include/vast/Util/DialectConversion.hpp:41]: mlir::ConversionPatternRewriter -> rewriter_t

File: include/vast/Util/DialectConversion.hpp [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    //             need to rethink how pieces fit together.
+    template< typename Op, typename TC, template< typename > class Impl >
+    struct TypeConvertingPattern : BasePattern< Op, Impl >
+    {
+        using parent_t = BasePattern< Op, Impl >;
+        using operation_t = typename parent_t::operation_t;
+
+        TC &tc;
+
+        TypeConvertingPattern(TC &tc, mlir::MLIRContext *mctx)
```
Comment (xlauko) [inline include/vast/Util/DialectConversion.hpp:67]: use MContext from Util/Common.hpp

File: include/vast/Util/DialectConversion.hpp [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+VAST_RELAX_WARNINGS
+#include <mlir/Transforms/DialectConversion.h>
+VAST_UNRELAX_WARNINGS
+
+namespace vast::util
+{
+    template< typename Op >
+    struct State
+    {
+        using rewriter_t = mlir::ConversionPatternRewriter;
```
Comment (xlauko) [inline include/vast/Util/DialectConversion.hpp:14]: I suggest to create some common usings for rewrite patterns at some shared place ... similar to Util/Common.hpp

File: include/vast/Util/DialectConversion.hpp [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        TC &tc;
+
+        TypeConvertingPattern(TC &tc, mlir::MLIRContext *mctx)
+            : parent_t(mctx), tc(tc)
+        {}
+
+        mlir::LogicalResult matchAndRewrite(
+                operation_t op,
+                typename operation_t::Adaptor ops,
+                mlir::ConversionPatternRewriter &rewriter) const override
```
Comment (xlauko) [inline include/vast/Util/DialectConversion.hpp:74]: rewriter_t

File: lib/vast/Conversion/Common/ToLLVM.cpp [hunk 11]
```diff
@@ -1,6 +1,6 @@
 // Copyright (c) 2021-present, Trail of Bits, Inc.
```
Comment (xlauko) [inline lib/vast/Conversion/Common/ToLLVM.cpp:1]: ToLLVM is a bit ambiguous name, what is a source of conversion?

File: lib/vast/Conversion/Common/ToLLVM.cpp [hunk 12]
```diff
@@ -51,226 +57,184 @@ namespace vast::hl
...
-        bool is_variadic(FuncOp op)
-        {
-            // TODO(lukas): Implement once hl supports it.
-            return false;
-        }
+        using TypeConverter = util::tc::LLVMTypeConverter;
 
-        auto convert_fn_t(auto &tc, FuncOp op)
-            -> std::tuple< mlir::TypeConverter::SignatureConversion, mlir::Type >
+        template< typename O >
```
Comment (xlauko) [inline lib/vast/Conversion/Common/ToLLVM.cpp:62]: use Op as at other places

**Review by xlauko — APPROVED** [review]

File: include/vast/Dialect/LowLevel/Passes.td (hunk 6 above)
Comment (lkorenc) [reply include/vast/Dialect/LowLevel/Passes.td:8]: Will do `CoreToLLVM` as this unfortunately requires the module to be in hybrid state of `ll + hl` right now.

**Pull request merged**